### PR TITLE
Revert #97980 Check for sandboxes before deleting the pod from apiserver

### LINKED ISSUE #88543: Kubelet deletes the pod before cleaning the CNI resources
**What would you like to be added**: I would like to make kubelet wait for CNI resources to be completely cleaned before deleting the pod from the apiserver. Currently it just waits for containers to be deleted and volumes to be unmounted.

**Why is this needed**:
There are couple of reason why this would make sense:
1) This will make sure that apiserver would provide the true view of the system. Because when pod is not cleaned from CNI because of any issue, then it's better for apiserver to show that the pod is still not terminated properly instead of cleaning up from etcd and relying on the GC to take care of it.
2) Currently the CNI's rely on the apiserver for additional metadata during creation and cleaning up since the kubelet doesn't pass additional metadata to the CNI which was discussed [here](https://github.com/kubernetes/kubernetes/issues/69882) and with apiserver deleting the pod before CNI is successful means CNI has to store kubernetes metadata separately.

### PR DESCRIPTION
**Pull request by kmala**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR is to commit #92817 back into master which was reverted because of performance issue related to deletion of pods. The reason for the performance issue was because during the check for sandbox deletion pod status is not correctly updated in the status manager. In order to fix the performance issue changed the code to now check whether the pod is deleted or not from the cache instead of status manager.

**Which issue(s) this PR fixes**:

Fixes #88543

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Users will see increase in time for deletion of pods and also guarantee that removal of pod from api server  would mean deletion of all the resources from container runtime.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
Have you tried this fix with the repro from https://github.com/kubernetes/kubernetes/issues/97288#issuecomment-756214750?

What I observed is that at the time of ContainerRemoved PLEG event, Pod is not yet marked for deletion. Is this work better to detect that Pod is "about to be deleted":

image

**Comment by kmala** [comment]
> Have you tried this fix with the repro from [#97288 (comment)](https://github.com/kubernetes/kubernetes/issues/97288#issuecomment-756214750)?
> 

Yes i had tried the reproduction scenario and could see that the pods are getting deleted quickly.

> What I observed is that at the time of ContainerRemoved PLEG event, Pod is not yet marked for deletion. Is this work better to detect that Pod is "about to be deleted":
> 
> image
> 
> ([commit](https://github.com/kubernetes/kubernetes/pull/98934/commits/c471431364d6ddd04329249fd614910df0408923))

I didn't understand this comment, can you please explain a bit more?

**Comment by kmala** [comment]
@SergeyKanzhelev can you please review when you get time

**Review by SergeyKanzhelev — COMMENTED** [review]
> > Have you tried this fix with the repro from [#97288 (comment)](https://github.com/kubernetes/kubernetes/issues/97288#issuecomment-756214750)?
> 
> Yes i had tried the reproduction scenario and could see that the pods are getting deleted quickly.

Can you confirm, you tried that scenario with 10 pods that was suggested in the issue discussion?

> > What I observed is that at the time of ContainerRemoved PLEG event, Pod is not yet marked for deletion. Is this work better to detect that Pod is "about to be deleted":
> > image
> > ([commit](https://github.com/kubernetes/kubernetes/pull/98934/commits/c471431364d6ddd04329249fd614910df0408923))
> 
> I didn't understand this comment, can you please explain a bit more?

What I found investigating the issue is that at the PLEG event "ContainerRemoved", the condition `kl.IsPodDeleted` returned `false` for some containers when many pods are being removed at the same time. So `toKeep` was set to `1` and sandbox is only deleted at GC.

In this PR, you moved the insides of `kl.IsPodDeleted` into the new location. But the condition is still the same - not evicted and deletion timestamp not nil while pod is not running. Unless I am missing something, this revert will fail the same way as the original PR. Thus the question, maybe I am missing how this condition was modified and how this code guarantees that at this PLEG event the Pod will be marked as not running already.

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go [hunk 1]
```diff
@@ -161,26 +161,13 @@ func (cgc *containerGC) removeOldestNSandboxes(sandboxes []sandboxGCInfo, toRemo
 	// Remove from oldest to newest (last to first).
 	for i := len(sandboxes) - 1; i >= numToKeep; i-- {
 		if !sandboxes[i].active {
-			cgc.removeSandbox(sandboxes[i].id)
+			if err := cgc.manager.DeleteSandbox(sandboxes[i].id); err != nil {
+				klog.ErrorS(err, "Failed to remove sandbox", "sandboxID", sandboxes[i].id)
```
[the remark this answers is not in the document]
Comment (kmala) [reply pkg/kubelet/kuberuntime/kuberuntime_gc.go:165]: The function doesn't have access to pod id though i can pass it as an argument didn't think to change the function signature just for logging. I can do it if you think it's needed

**Comment by kmala** [comment]
> > > Have you tried this fix with the repro from [#97288 (comment)](https://github.com/kubernetes/kubernetes/issues/97288#issuecomment-756214750)?
> > 
> > 
> > Yes i had tried the reproduction scenario and could see that the pods are getting deleted quickly.
> 
> Can you confirm, you tried that scenario with 10 pods that was suggested in the issue discussion?

  Yes, i had created replicaset with 10 pods on kind cluster and then deleted the replica set.

> > > What I observed is that at the time of ContainerRemoved PLEG event, Pod is not yet marked for deletion. Is this work better to detect that Pod is "about to be deleted":
> > > image
> > > ([commit](https://github.com/kubernetes/kubernetes/pull/98934/commits/c471431364d6ddd04329249fd614910df0408923))
> > 
> > 
> > I didn't understand this comment, can you please explain a bit more?
> 
> What I found investigating the issue is that at the PLEG event "ContainerRemoved", the condition `kl.IsPodDeleted` returned `false` for some containers when many pods are being removed at the same time. So `toKeep` was set to `1` and sandbox is only deleted at GC.
> 
> In this PR, you moved the insides of `kl.IsPodDeleted` into the new location. But the condition is still the same - not evicted and deletion timestamp not nil while pod is not running. Unless I am missing something, this revert will fail the same way as the original PR. Thus the question, maybe I am missing how this condition was modified and how this code guarantees that at this PLEG event the Pod will be marked as not running already.

The reason it wasn't working before was `kl.IsPodDeleted` gets the pod status from status manager which is not updated always during the PLEG event handling time and i changed the code to get the pod status from cache instead where it has the latest status always/

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmala) [reply pkg/kubelet/kuberuntime/kuberuntime_gc.go:165]: fixed it

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go [hunk 2]
```diff
@@ -309,3 +309,15 @@ func (m *kubeGenericRuntimeManager) GetPortForward(podName, podNamespace string,
 	}
 	return url.Parse(resp.Url)
 }
+
+// DeleteSandbox removes the sandbox by sandboxID.
+func (m *kubeGenericRuntimeManager) DeleteSandbox(sandboxID string) error {
+	klog.V(4).InfoS("Removing sandbox", "sandboxId", sandboxID)
```
[the remark this answers is not in the document]
Comment (kmala) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:315]: changed the method name

**Comment by kmala** [comment]
@SergeyKanzhelev  can you please approve again, i just did a rebase

**Comment by kmala** [comment]
@sjenning @rphillips can you please take look when you get chance

**Comment by smarterclayton** [comment]
/hold

I want to look into this, it may have been addressed by #102344

File: pkg/kubelet/kuberuntime/kuberuntime_gc.go [hunk 3]
```diff
@@ -161,26 +161,13 @@ func (cgc *containerGC) removeOldestNSandboxes(sandboxes []sandboxGCInfo, toRemo
...
-			cgc.removeSandbox(sandboxes[i].id)
+			if err := cgc.manager.RemoveSandbox(sandboxes[i].id); err != nil {
+				klog.ErrorS(err, "Failed to remove sandbox", "sandboxID", sandboxes[i].id)
+			}
 		}
 	}
 }
 
-// removeSandbox removes the sandbox by sandboxID.
-func (cgc *containerGC) removeSandbox(sandboxID string) {
```
Comment (smarterclayton) [inline pkg/kubelet/kuberuntime/kuberuntime_gc.go:170]: Why isn't this change just to make `removeSandbox` here return an error?  That makes this a much smaller change (you could just change containerGC and make this a 4-10 line PR instead of 400).

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1980,6 +1984,9 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 				klog.V(4).InfoS("SyncLoop (PLEG): pod does not exist, ignore irrelevant event", "event", e)
 			}
 		}
+		if e.Type == pleg.ContainerRemoved {
+			kl.deletePodSandbox(e.ID)
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet.go:1988]: This no longer belongs here, this belongs in pod worker `syncTerminatedPod`.  As a result of the changes in #102344 I'm pretty sure it is already there, so this is an unnecessary change and should be removed.

**Comment by smarterclayton** [comment]
/hold

Added some review comments:

1. The changes in containerGC seem like they should be much simpler (returning an error in the removeSandbox method)
2. The changes in kubelet.go should be unnecessary after #102344 - sandboxes are repeatedly cleaned up by the pod worker until they are gone and we don't want separate loops for that in Kubelet

This should not be merged as is because of 2, although 1 needs to be answered in case I'm missing something.

File: pkg/kubelet/kubelet.go (hunk 4 above)
Comment (kmala) [reply pkg/kubelet/kubelet.go:1988]: I looked into the PR and it doesn't seem to fix this scenario atleast from the looks of it. I can test it once to confirm. The reason why i don't feel the PR fixes is because in the syncTerminatedPod we call KillPod which only stops the sandboxes but doesn't remove them and leaves it to the GC to do that.

**Comment by kmala** [comment]
> /hold
> 
> Added some review comments:
> 
> 1. The changes in containerGC seem like they should be much simpler (returning an error in the removeSandbox method)
> 2. The changes in kubelet.go should be unnecessary after [Prevent Kubelet from incorrectly interpreting "not yet started" pods as "ready to terminate pods" by unifying responsibility for pod lifecycle into pod worker #102344](https://github.com/kubernetes/kubernetes/pull/102344) - sandboxes are repeatedly cleaned up by the pod worker until they are gone and we don't need separate loops for that in Kubelet anymore (I believe)
> 
> This should not be merged as is because of 2, although 1 needs to be answered in case I'm missing something.

I think if i explain from the start of the issue, it would clarify more:
1) we were facing an issue where pod is deleted from the k8s apiserver even though all the sandboxes are not cleaned up properly(cni in our case) for which i created the issue https://github.com/kubernetes/kubernetes/issues/88543
2) I had raised a [PR](https://github.com/kubernetes/kubernetes/pull/89667/files) to fix the issue with a simple code change to check to wait for all the sandboxes to removed similar to how we wait for the containers in [PodResourcesAreReclaimed](https://github.com/kubernetes/kubernetes/blob/9f47110aa29094ed2878cf1d85874cb59214664a/pkg/kubelet/kubelet_pods.go#L925) function.
3) This was reverted during the release testing because it is causing test cases to take long time to execute. The reason for the performance issue was because sandboxes are removed only during GC and hence waiting for the removal of sandboxes to remove the pod will be increased. In order to fix the performance issue added a new sandbox deleter which is similar to container deleter which gets triggered upon container removal event. Merged this change through the [PR](https://github.com/kubernetes/kubernetes/pull/92817) but was also reverted because of another scale issue which happened because status manager doesn't have the latest change.  **This is the reason for the PR being this big.**

**Comment by smarterclayton** [comment]
Ok, given that, the changes in #102344 make it so that the best place to synchronously clean up resources is syncTerminatedPod.  If you want to add something to PodResourcesAreReclaimed, you either wait for it to GC (not ideal) or you add synchronous cleanup code to syncTerminatedPod.

However, we can't remove containers until pods are deleted in order to preserve container logs until the pod is removed between dockershim (which requires containers to be present in order to view logs) and CRI https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1845. We decided that until we remove dockershim we can't remove containers until after deletion, which also means we can't remove sandboxes.  So it would not be appropriate to add sandbox deletion to syncTerminatedPod unless we're willing to break that contract.

@derekwaynecarr  @ehashman @Random-Liu I think we need to decide / formalize how resource cleanup for containers and sandboxes are managed (and what guarantees are provided for users, such as being able to view pod logs until the api object is fully deleted) before we go forward with this change.  I will admit that I did not find the argument in #88543 compelling enough to break pod log availability now, but the direction (delete what we can before full termination) is understandable.  Dawn seemed to be much more in favor of reducing latency for full delete, and I am as well, but any resource we cleanup after etcd termination reduces the kubelet's ability to manage workload cost.

**Comment by kmala** [comment]
From what i understood, i see we don't want to wait for sandbox cleanup to remove the pod object from the etcd. If we are saying this then wouldn't it be against the correctness of the system/cluster? The apiserver doesn't know of the pod but the pod is still consuming resources (here network) from the cluster? We wait for the cgroups and volumes to be cleaned up before removing the pod from api server, then why not wait for the network resources also to be cleaned up?
I totally understand why the current logic is the way it is and your point on the logs but i feel the system is not behaving correctly if we are leaving so many things for the GC ( the reason i say many things is because since the CNI can be chained , and they need to have separate metadata instead of relying on the kubernetes for reconciliation)and the only other way i see is that for the CRI to inform the network resources cleanup, so that containers can remain for the GC but the important resources are cleanup.

**Comment by smarterclayton** [comment]
> The apiserver doesn't know of the pod but the pod is still consuming resources (here network) from the cluster? We wait for the cgroups and volumes to be cleaned up before removing the pod from api server, then why not wait for the network resources also to be cleaned up?

Currently, we maintain containers to preserve logs, so we can't remove network resources until we remove containers.  A change that wanted to remove network resources should probably start by working to separate log behavior from container behavior, then we could begin tearing those down prior to pod deletion.  But that would need more work.

As far as CNI correctness, there is nothing conceptually that prevents the Kubelet from instructing the CNI to release a pod once it passes into syncTerminatedPod, but that is probably a more involved change and probably should wait until we decide when to change cleanup.

> If we are saying this then wouldn't it be against the correctness of the system/cluster? The apiserver doesn't know of the pod but the pod is still consuming resources (here network) from the cluster? We wait for the cgroups and volumes to be cleaned up before removing the pod from api server, then why not wait for the network resources also to be cleaned up?

A pod can be force deleted at any time (although we discourage it) and so no - it's not a violation of system correctness to remove some types pod resources after the object no longer exists in etcd.  Volume cleanup is only related to correctness in that we have provided the guarantee that certain types of volumes are only mounted to one node at a time with running processes accessing it (and so we wait for all processes to be stopped before we allow teardown of the pod, and a force delete is a user overriding that and accepting the side effects).

**Comment by kmala** [comment]
> > The apiserver doesn't know of the pod but the pod is still consuming resources (here network) from the cluster? We wait for the cgroups and volumes to be cleaned up before removing the pod from api server, then why not wait for the network resources also to be cleaned up?
> 
> Currently, we maintain containers to preserve logs, so we can't remove network resources until we remove containers. A change that wanted to remove network resources should probably start by working to separate log behavior from container behavior, then we could begin tearing those down prior to pod deletion. But that would need more work.
> As far as CNI correctness, there is nothing conceptually that prevents the Kubelet from instructing the CNI to release a pod once it passes into syncTerminatedPod, but that is probably a more involved change and probably should wait until we decide when to change cleanup.

Yes, i understand this cyclic dependency where we need logs before we delete pod from etcd , for logs we need containers and we can't delete network unless containers are deleted. i am open to discussing on this separation of logs and containers.
Yes, kubelet making CNI calls is going to be big and also it would bring the same point as above can we tear down network before removing the containers?

> > If we are saying this then wouldn't it be against the correctness of the system/cluster? The apiserver doesn't know of the pod but the pod is still consuming resources (here network) from the cluster? We wait for the cgroups and volumes to be cleaned up before removing the pod from api server, then why not wait for the network resources also to be cleaned up?
> 
> A pod can be force deleted at any time (although we discourage it) and so no - it's not a violation of system correctness to remove some types pod resources after the object no longer exists in etcd. Volume cleanup is only related to correctness in that we have provided the guarantee that certain types of volumes are only mounted to one node at a time with running processes accessing it (and so we wait for all processes to be stopped before we allow teardown of the pod, and a force delete is a user overriding that and accepting the side effects).

Yes, when user is doing a force delete, we don't guarantee anything but when the pod is gracefully shutdown which is the case i am referring to i see it as an issue with system correctness.  In the graceful shutdown scenario, we wait for the volumes and cgroups to be properly cleaned up before deleting the pod from etcd and i feel the same should apply for network.


I can add this to the SIG Node meeting agenda so that we can discuss more on it instead of chat as it would be more easier and faster.

**Comment by smarterclayton** [comment]
> Yes, when user is doing a force delete, we don't guarantee anything but when the pod is gracefully shutdown which is the case i am referring to i see it as an issue with system correctness

When a user does a force delete correctly (when they force delete a pod on a node they know is shut down) we do preserve correctness.  Etcd object deletion of a pod in Kube is defined to mean "no processes are running that can access the pod's resources" whether triggered by kubelet or by an admin.  Any network plugin should be able to make the same assumption, and I would be concerned if CNI plugins are assuming that they will always get CNI remove calls (not guaranteed at all).  Before we meet in sig-node, please describe what you mean by "system correctness" (since that definition is what you are using to argue here).  There are other advantages to serial cleanup like latency reduction and backpressure that can be solved in other ways.

**Comment by kmala** [comment]
> > Yes, when user is doing a force delete, we don't guarantee anything but when the pod is gracefully shutdown which is the case i am referring to i see it as an issue with system correctness
> 
> When a user does a force delete correctly (when they force delete a pod on a node they know is shut down) we do preserve correctness. Etcd object deletion of a pod in Kube is defined to mean "no processes are running that can access the pod's resources" whether triggered by kubelet or by an admin. Any network plugin should be able to make the same assumption, and I would be concerned if CNI plugins are assuming that they will always get CNI remove calls (not guaranteed at all). Before we meet in sig-node, please describe what you mean by "system correctness" (since that definition is what you are using to argue here). There are other advantages to serial cleanup like latency reduction and backpressure that can be solved in other ways.

I am surprised to hear that CNI remove calls are not guaranteed at all. The reason i am saying system correctness is because similar to memory, cpu and volumes, network is also a resource consumed by the pod and is active as long as its cleaned properly and merely stopping of containers won't release that resource. If we currently see after a pod is removed from the etcd currently network is the only active resource used we wait for the containers to be killed and removal of all other resources like cgroups and volumes . So, i feel we have to consider this a leak as there are no pods but there are network resources consumed in the cluster because there is very much a probability of not being able to launch a pod because previous pods network resources are not properly cleaned up. 
The reason i was asking about discussion in SIG-Node is because i had discussed the issue and fix before in sig node and got approval before creating a PR and hence wanted to see what everyone feels now and also if we want to check if anyone has different ideas on how we can implement the requirement.

**Comment by Random-Liu** [comment]
> From what i understood, i see we don't want to wait for sandbox cleanup to remove the pod object from the etcd. If we are saying this then wouldn't it be against the correctness of the system/cluster? The apiserver doesn't know of the pod but the pod is still consuming resources (here network) from the cluster? We wait for the cgroups and volumes to be cleaned up before removing the pod from api server, then why not wait for the network resources also to be cleaned up?

I believe we try to guarantee that the pod sandbox is successfully stopped once before deleting the pod. However, there is probably no way to guarantee that, e.g. a pod is forcibly deleted, or a pod is deleted when kubelet is down.

However, even in that case I believe kubelet will stop the pod sandbox once before removing it. The problem is that the CNI plugin should tolerate that scenario, and properly cleanup the pod in that scenario. I believe that is also why we built checkpoint into the dockershim https://github.com/kubernetes/kubernetes/commit/51526d310385d6f3791db8bd00ed16abab46658f

**Comment by kmala** [comment]
closing as the issue is fixed with https://github.com/kubernetes/kubernetes/pull/102344
